### ISSUE TITLE
False-positive UnusedVariable when assigning to references

### ISSUE THREAD

**Issue by pilif** [vimeo/psalm#4472]
https://psalm.dev/r/0649bdda79

this was fine in 3.15 but is marked as an error in 4.1. 

While this reduced sample would allow for the argument that the assignment to $l is not strictly needed, the original code is part of a more complicated class implementing php_user_filter where that variable is being used after the bucket where it was taken from was being written to.

The more complicated case where that argument certainly doesn't hold:

https://psalm.dev/r/9e73bac9ca

IMHO, even in the simpler case, the the assignment is valid because the variable certainly *is* read.

Also, I agree that the code is nasty with the reference, but given the use case (user filter) the signature is out of my control.

**Comment by pilif** [comment]
in our somewhat aged code-base, we have around 20 instances of functions writing to a parameter passed by reference. Those are all affected by this issue, causing variables to be flagged as unused even though they are used.

While I can probably fix most of them to return the thing they are currently writing into an reference parameter, but I still think this is worth fixing.

**Comment by muglug** [comment]
Similar issue with

```php
<?php

/** @param array<array> $data */
function foo(array $data) : array {
    $output_rows = [];
    array_map(
        function (array $row) use (&$output_rows) {
            $output_rows[] = $row;
        },
        $data
    );

    return $output_rows;
}
```
